src/Dom/Document.php

```diff
@@ -108,6 +108,13 @@ final class Document extends DOMDocument
      */
     const XPATH_INLINE_STYLE_ATTRIBUTES_QUERY = './/@style';
 
+    /**
+     * Associative array for lazily-created, cached properties for the document.
+     *
+     * @var array
+     */
+    private $properties = [];
+
```

Comment:
I just realized a potential large simplification here. Why if each of the properties were instead just declared as null?

```suggestion
    public $xpath
    public $html;
    public $head;
    public $body;
    public $charset;
    public $viewport;
    public $ampElements;
    public $ampCustomStyle;
    public $ampCustomStyleByteCount;
    public $inlineStyleByteCount
    public $links;
```

The phpdoc not withstanding.

Reply:
I see you did this in 525e91853ce74d56cdd656ee08dd5494ed3ad200 but then reverted. Did you try `private` instead of `public`? This would model what is done with the `$links` property already.

Reply:
I see the problem with that approach. It's that private access in the class won't cause the getter to be invoked. I think I have a way around that, and that is to manually invoke `__get()` for methods on `Document`. I'm going to give it a try, though the commit may not be pushed until Monday.

Reply:
Bah. Doing this makes PHPStan blow up with a lot of "Access to private property" errors.